crates/mako/src/plugins/duplicate_package_checker.rs

```diff
@@ -0,0 +1,167 @@
+fn clean_path(module_path: &Path) -> PathBuf {
+    let path_str = module_path.to_string_lossy();
+    let cleaned_path = path_str
+        .replace("/node_modules/", "/~/")
+        .replace("\\node_modules\\", "/~/");
+    PathBuf::from(cleaned_path)
+}
+
+/// Makes the cleaned path relative to the given context
+fn clean_path_relative_to_context(module_path: &Path, context: &Path) -> PathBuf {
+    let cleaned_path = clean_path(module_path);
+    let context_str = context.to_str().unwrap();
+    let cleaned_path_str = cleaned_path.to_str().unwrap();
+
+    if cleaned_path_str.starts_with(context_str) {
+        let relative_path = cleaned_path_str.trim_start_matches(context_str);
+        PathBuf::from(format!(".{}", relative_path))
+    } else {
+        cleaned_path
+    }
+}
+
+impl DuplicatePackageCheckerPlugin {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn verbose(mut self, verbose: bool) -> Self {
+        self.verbose = verbose;
+        self
+    }
+
+    pub fn show_help(mut self, show_help: bool) -> Self {
+        self.show_help = show_help;
+        self
+    }
+
+    pub fn emit_error(mut self, emit_error: bool) -> Self {
+        self.emit_error = emit_error;
+        self
+    }
+
+    fn find_duplicates(packages: Vec<PackageInfo>) -> HashMap<String, Vec<PackageInfo>> {
+        let mut package_map: HashMap<String, Vec<PackageInfo>> = HashMap::new();
+
+        for package in packages {
+            package_map
+                .entry(package.name.clone())
+                .or_default()
+                .push(package);
+        }
+
+        package_map
+            .into_iter()
+            .filter(|(_, versions)| versions.len() > 1)
+            .collect()
+    }
+
+    fn check_duplicates(
+        &self,
+        module_graph: &RwLock<ModuleGraph>,
+    ) -> HashMap<String, Vec<PackageInfo>> {
+        let mut packages = Vec::new();
+
+        module_graph
+            .read()
+            .unwrap()
+            .modules()
+            .iter()
+            .for_each(|module| {
+                if let Some(ResolverResource::Resolved(resource)) = module
+                    .info
+                    .as_ref()
+                    .and_then(|info| info.resolved_resource.as_ref())
+                {
+                    let package_json = resource.0.package_json().unwrap();
+                    let raw_json = package_json.raw_json();
+                    let default_version = serde_json::Value::String("0.0.0".to_string());
+                    let version = raw_json
+                        .as_object()
+                        .unwrap()
+                        .get("version")
+                        .unwrap_or(&default_version)
+                        .as_str()
+                        .unwrap();
+                    let package_info = PackageInfo {
+                        name: package_json.name.clone().unwrap_or("mako-pkg".to_string()),
```

Comment:
what about  ignoring  package.json without a name?

if there two "unnamed" package.json,  the plugin will  treat them as duplicated with a confusing name "mako-pkg"

Reply:
Oh, thanks for the heads-up, I'll fix it right now.

Reply:
> what about ignoring package.json without a name?
> 
> if there two "unnamed" package.json, the plugin will treat them as duplicated with a confusing name "mako-pkg"

Use `if let` to process the `package.json` if it's missing a version or name, please review again. @stormslowly